src/merkle/smt/mod.rs

```diff
@@ -244,3 +323,237 @@ impl<const DEPTH: u8> TryFrom<NodeIndex> for LeafIndex<DEPTH> {
+
+        use NodeMutation::*;
+
+        match self.node_mutations.entry(index) {
+            Vacant(entry) => {
+                entry.insert(Addition(new_node));
+            },
+            Occupied(mut entry) => match entry.get_mut() {
+                Addition(existing) => {
+                    // If there's an existing addition with this key, then overwrite it to be an
+                    // addition of this new node instead.
+                    *existing = new_node;
+                },
+                Removal => {
+                    // Likewise a removal of this key gets overwritten with an addition.
+                    entry.insert(Addition(new_node));
+                },
+            },
+        }
+    }
+}
+
+/// Helper type that represents a [`MutationSet`] that's ready to be applied to a
+/// [`SparseMerkleTree`].
+///
+/// It is created by [`MutationSet::done()`], which should be directly passed to
+/// [`SparseMerkleTree::commit()`]:
+/// ```
+/// # use miden_crypto::{hash::rpo::RpoDigest, merkle::Smt, Felt};
+/// # let mut smt = Smt::default();
+/// # let key = RpoDigest::default();
+/// # let value: [Felt; 4] = Default::default();
+/// let mut mutations = smt.mutate();
+/// mutations.insert(key, value);
+/// smt.commit(mutations.done());
+/// ```
+// This type exists for the sake of the borrow checker -- SparseMerkleTree::commit() needs a
+// mutable reference to `self`, but MutationSet stores a shared reference to the SparseMerkleTree,
+// and those can't both exist at the same time. By going through this type first, which stores all
+// the same data as MutationSet except for the shared reference to SparseMerkleTree, the shared
+// reference is dropped just before SparseMerkleTree::commit() is called, and the borrow checker is
+// happy. Interior mutability would also work, but then we would lose static lifetime checking.
+pub struct CompletedMutationSet<const DEPTH: u8, K, V> {
+    /// Corresponds to MutationSet's `node_mutations` field, and constructed from it.
+    ///
+    /// This is the set of nodes that need to be removed or added. Each [`NodeMutation::Addition`]
+    /// corresponds to a [`SparseMerkleTree::add_inner_node()`] call, and each
+    /// [`NodeMutation::Removal`] corresponds to a [`SparseMerkleTree::remove_inner_node()`] call.
+    node_mutations: BTreeMap<NodeIndex, NodeMutation>,
+    /// Corresponds to MutationSet's `new_pairs` field, and constructed from it.
+    ///
+    /// This is the set of key-value pairs that need to be inserted. Each entry corresponds to a
+    /// [`SparseMerkleTree::insert_value()`] call.
+    new_pairs: BTreeMap<K, V>,
+    /// Corresponds to MutationSet's `new_root` field, and constructed from it.
+    ///
+    /// This is the Merkle tree's root hash that was calculated on the last
+    /// [`MutationSet::insert()`] call, and is applied to the Merkle tree with
+    /// [`SparseMerkleTree::set_root()`].
+    new_root: RpoDigest,
```

Comment:
Here, if we also add `old_root`, we should be able to tell if a given mutation set can be applied to a given tree.

Reply:
Good idea!